arangod/Aql/OptimizerRulePropagateConstantAttributes.cpp

```diff
@@ -0,0 +1,344 @@
+#include "Logger/LogMacros.h"
+#include "OptimizerRules.h"
+#include "OptimizerUtils.h"
+#include "VocBase/LogicalCollection.h"
+
+#include "Containers/ImmutableMap.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using namespace arangodb::containers;
+using EN = arangodb::aql::ExecutionNode;
+
+#define LOG_RULE LOG_DEVEL_IF(false)
+
+namespace {
+
+struct Replacement {
+  std::uint64_t limitCount;
+  AstNode const* ast;
+};
+
+struct VarAttributeAccess {
+  Variable const* var;
+  containers::SmallVector<std::string_view, 8> path;
+
+  friend bool operator==(VarAttributeAccess const&,
+                         VarAttributeAccess const&) noexcept = default;
+
+  friend std::size_t hash_value(VarAttributeAccess const& key) {
+    std::size_t seed = 0;
+    boost::hash_combine(seed, key.var);
+    boost::hash_range(seed, key.path.begin(), key.path.end());
+    return seed;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os,
+                                  VarAttributeAccess const& x) {
+    os << x.var->name << " [";
+    std::copy(x.path.begin(), x.path.end(),
+              std::ostream_iterator<std::string_view>(os, ", "));
+    return os << "]";
+  }
+};
+
+using VarReplacementMap =
+    containers::ImmutableMap<VarAttributeAccess, Replacement,
+                             boost::hash<VarAttributeAccess>>;
+using VarReplacementTransientMap =
+    containers::ImmutableMapTransient<VarAttributeAccess, Replacement,
+                                      boost::hash<VarAttributeAccess>>;
+
+/// @brief extract an attribute and its variable from an attribute access
+/// (e.g. `a.b.c` will return variable `a` and attribute name `b.c.`.
+bool getAttribute(AstNode const* attribute,
+                  VarAttributeAccess& attributeAccess) {
+  TRI_ASSERT(attribute != nullptr);
+  if (attribute->type == NODE_TYPE_REFERENCE) {
+    attributeAccess.var = static_cast<Variable const*>(attribute->getData());
+    return true;
+  } else if (attribute->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
+    getAttribute(attribute->getMember(0), attributeAccess);
+    attributeAccess.path.emplace_back(attribute->getStringView());
+    return true;
+  }
+  return false;
+}
+
+/// @brief inspect the constant value assigned to an attribute
+/// the attribute value will be stored so it can be inserted for the attribute
+/// later
+void inspectConstantAttribute(AstNode const* attribute, AstNode const* value,
+                              std::size_t currentLimitCount,
+                              VarReplacementTransientMap& replacements) {
+  VarAttributeAccess access;
+  LOG_RULE << "ATTR = " << attribute->toString();
+  if (!getAttribute(attribute, access)) {
+    return;
+  }
+  LOG_RULE << "VAR = " << access;
+  auto it = replacements.find(access);
+
+  if (it == nullptr) {
+    replacements.set(access, Replacement{currentLimitCount, value});
+    return;
+  }
+}
+
+void collectConstantAttributes(AstNode* node, std::size_t currentLimitCount,
```

Comment:
I hope that we can make `AstNode* node` const here, so we can avoid the const_cast further down below at the call site of this function.